Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,664 @@
+    public bool Started { get; set; } = false;
+    public bool Finished { get; set; } = false;
+    public DateTime StartTime { get; protected set; }
+    public DateTime EndTime { get; protected set; }
+    public int TimeSpent => (int)(Finished ? ((TimeSpan)(EndTime - StartTime)).TotalMilliseconds : Started ? (DateTime.Now - StartTime).TotalMilliseconds : 0);
+    // Resources allocated at the end of the computation.
+    public int ResourceCount { get; set; } = 0;
+
+
+
+    // If this node is an error, all the trace positions
+    public ImmutableList<Range> RelatedRanges { get; set; } = ImmutableList<Range>.Empty;
+
+    // Sub-diagnostics if any
+    public List<NodeDiagnostic> Children { get; set; } = new();
+    private List<NodeDiagnostic> NewChildren { get; set; } = new();
+
+    public int GetNewChildrenCount() {
+      return NewChildren.Count;
+    }
+
+    public void AddNewChild(NodeDiagnostic newChild) {
+      NewChildren.Add(newChild);
+    }
+
+    public void SaveNewChildren() {
+      Children = NewChildren;
+      ResetNewChildren();
+    }
+    public void ResetNewChildren() {
+      NewChildren = new();
+    }
+
+    public NodeDiagnostic SetObsolete() {
+      if (StatusCurrent != CurrentStatus.Obsolete) {
+        StatusCurrent = CurrentStatus.Obsolete;
+        foreach (var child in Children) {
+          child.SetObsolete();
+        }
+      }
+
+      return this;
+    }
+
+    // Returns true if it started the method, false if it was already started
+    public virtual bool Start() {
+      if (StatusCurrent != CurrentStatus.Verifying || !Started) {
+        StartTime = DateTime.Now;
+        StatusCurrent = CurrentStatus.Verifying;
+        foreach (var child in Children) {
+          child.Start();
+        }
+        Started = true;
+        return true;
+      }
+
+      return false;
+    }
+
+    // Returns true if it did stop the current node, false if it was already stopped
+    public virtual bool Stop() {
+      if (StatusCurrent != CurrentStatus.Current || !Finished) {
+        EndTime = DateTime.Now;
+        StatusCurrent = CurrentStatus.Current;
+        foreach (var child in Children) {
+          child.Stop();
+        }
+        Finished = true;
+        return true;
+      }
+
+      return false;
+    }
+
+    public void PropagateChildrenErrorsUp() {
+      var childrenHaveErrors = false;
+      foreach (var child in Children) {
+        child.PropagateChildrenErrorsUp();
+        if (child.StatusVerification == VerificationStatus.Error) {
+          childrenHaveErrors = true;
+        }
+      }
+
+      if (childrenHaveErrors) {
+        StatusVerification = VerificationStatus.Error;
+      }
+    }
+
+    // Requires PropagateChildrenErrorsUp to have been called before.
+    public virtual void RenderInto(LineVerificationStatus[] perLineDiagnostics, bool contextHasErrors = false, bool contextIsPending = false, Range? otherRange = null, Range? contextRange = null) {
```

Comment:
This is a quite complex yet straightforward way to convert the verification status and current status of a node into a "linear diagnostic", using double nested switch statements.
It is context-sensitive, so that if a node has errors, we display verifying assertions with special icons and not green lines.
Moreover, for ranges that exceed one line (i.e. everything not an assertion), we use special range icons to render them if they are errors.